plugins/ccgx/fu-ccgx-dmc-device.c

```diff
@@ -0,0 +1,703 @@
+		self->update_model = DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER;
+	else if (dmc_int_rqt.data[0] == DMC_DEVICE_STATUS_FW_DOWNLOADED_UPDATE_PEND)
+		self->update_model = DMC_UPDATE_MODEL_PENDING_RESET;
+	else {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid  status code = %u",
+			     dmc_int_rqt.data[0]);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static FuFirmware *
+fu_ccgx_dmc_device_prepare_firmware (FuDevice *device,
+				     GBytes *fw,
+				     FwupdInstallFlags flags,
+				     GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_ccgx_dmc_firmware_new ();
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	GBytes* custom_meta_record = NULL;
+	gboolean custom_meta_exist = FALSE;
+
+	/* parse all images */
+	if (!fu_firmware_parse (firmware, fw, flags, error))
+		return NULL;
+
+	/* get custom meta record */
+	custom_meta_record = fu_ccgx_dmc_firmware_get_custom_meta_record (FU_CCGX_DMC_FIRMWARE(firmware));
+	if (custom_meta_record)
+		if (g_bytes_get_size (custom_meta_record) > 0)
+			custom_meta_exist = TRUE;
+
+	/* check custom meta flag */
+	if ( self->dock_id.custom_meta_data_flag  !=  custom_meta_exist)  {
+		g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "custom meta data mismatch");
+		return NULL;
+	}
+	return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_ccgx_dmc_device_attach (FuDevice *device, GError **error)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	gboolean manual_replug = FALSE;
+	guint64 remove_delay_ms = 0;
+
+	if (fu_device_has_custom_flag (device, "has-manual-replug"))
+		manual_replug = TRUE;
+
+	if (fu_device_get_update_state (self) != FWUPD_UPDATE_STATE_SUCCESS)
+		return TRUE;
+
+	if (self->update_model == DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER) {
+		DmcTriggerCode trigger_code = DMC_TRIGGER_CODE_UPDATE_NOW;
+
+		if (manual_replug)
+			trigger_code = DMC_TRIGGER_CODE_UPDATE_ON_DISCONNECT;
+
+		g_debug ("download tirgger:%u",trigger_code);
+
+		if (!fu_ccgx_dmc_device_send_download_trigger (self,
+							       trigger_code,
+							       error)) {
+			g_prefix_error (error, "download trigger error:");
+			return FALSE;
+		}
+	} else if (self->update_model == DMC_UPDATE_MODEL_PENDING_RESET) {
+
+		g_debug ("pending reset:%u",(guint32)manual_replug);
+
+		if (!fu_ccgx_dmc_device_send_sort_reset (self,
+							       manual_replug,
+							       error)) {
+			g_prefix_error (error, "soft reset error:");
+			return FALSE;
+		}
+	}  else {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "invalid update model = %u",
+				     self->update_model);
+			return FALSE;
+	}
+
+	if (manual_replug)
+		return TRUE;
+
+	g_debug ("reset and wait replug");
+
+	remove_delay_ms = fu_device_get_install_duration(self) * 1000;
+	fu_device_set_remove_delay (FU_DEVICE (self), remove_delay_ms); /* over a minute to flash offline */
```

Comment:
is the replug delay really proportional to the amount of time taken to flash?

Reply:
Add InstallDuration and RemoveDelay on ccgx.quirk file on today commit (ccgx: fix code alignment, modify code for review).
Could you check whether RemoveDelay is right usage?